### DISCUSSION TITLE
How To split >2GB ONNX model with external data?

### DISCUSSION THREAD

**Discussion by lukolszewski** [onnx/onnx#5407] [Q&A] [answered]
I have a large (substantially larger than 2GB) ONNX model with external data files. I've been having trouble with what I thought would be a pretty simple task. Slice a part of that model at a node. The resulting "small" model is also over 2GB BTW.

Does anyone know any way this can be done without converting .onnx binary file format to txt with Protocol Buffers and then spending ages editing the resulting txt file in an editor by hand?

I've tried onnx.util.extract_model, it fails with:
ValueError: Message onnx.ModelProto exceeds maximum protobuf size of 2GB: 28869931499

I tried with check_model = False with same result.

I tried a python tool that has been written to bypass the 2GB limitation of onnx.util.extract_model named sne4onnx, but it fails with the exact same error message. I suspect it would allow me to work with models over 2GB, but the chunk being sliced has to be under 2GB. 

So, how do people split large (substantially larger than 2GB) ONNX models?

**Comment by gramalingam** [comment]
Did you try converting the model data to be external, as described [here](https://github.com/onnx/onnx/blob/main/docs/PythonAPIOverview.md#converting-an-onnx-model-to-external-data) ?

**Comment by kartikpodugu** [comment]
I want answer for this question too. I have a model which is more than 10GB. How to break the model into multiple parts?
I have one file named **model.onnx** and other file named **model.onnx_data**

**Reply by ALZ112** [reply]
Any updates on how did you solve the issue? me facing same error.

**Comment by ALZ112** [comment]
Guys Anyone solved the issue?

**Comment by justinchuby** [comment]
For new readers, here is a pass that does it https://github.com/microsoft/onnxscript/pull/2119. Combined with `onnxscript.ir.load` and `onnxscript.ir.save` you may work with large models without limitation.

**Reply by ochougul** [reply]
@justinchuby 
This does not work with graph that has subfunctions in it.
Which can be produced by passing `use_submodules_as_functions={pytorch module class name}` in `torch.onnx.export` call
Throws following error
```
Traceback (most recent call last):                                                                                       
  File "extract_graph.py", line 12, in <module>
    result = exg(model)                                                                                                  
  File "/site-packages/onnx_ir/passes/_pass_infra.py", line 127, in __call__
    result = self.call(model)
  File "onnx_extract_model.py", line 102, in call
    model.graph = ir.Graph(               
  File "site-packages/onnx_ir/_core.py", line 2881, in __init__
    self._initializers = _graph_containers.GraphInitializers(
  File "site-packages/onnx_ir/_graph_containers.py", line 230, in __init__
    self._set_graph(value)
  File "site-packages/onnx_ir/_graph_containers.py", line 237, in _set_graph
    raise ValueError(
ValueError: Value '%"onnx::MatMul_2018"<FLOAT,[12288,5120]>{ExternalTensor(...)}' is already an initializer of a different graph. Please remove the value from the previous graph first
```

setup:
```
onnx==1.18.0
onnx-ir==0.1.14
onnxruntime==1.22.0
onnxscript==0.5.7
torch==2.7.0+cpu
```

**Reply by justinchuby** [reply, marked as answer]
Can you test with this new method extract? https://onnx.ai/ir-py/api/ir_convenience.html#onnx_ir.convenience.extract

**Reply by justinchuby** [reply]
@ochougul if possible it would be helpful if you can share an onnx model, thanks!

**Reply by ochougul** [reply]
I tried with above, it works now! Thanks @justinchuby
